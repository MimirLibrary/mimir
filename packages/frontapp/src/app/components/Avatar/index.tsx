import { FC } from 'react';
import styled from '@emotion/styled';

const UserAvatar = styled.img`
  max-height: 7rem;
  max-width: 7rem;
  width: 100%;
  height: 100%;
  border-radius: 50%;
`;

interface IProps {
  src: string;
}

const Avatar: FC<IProps> = ({ src = '' }) => {
  return (
    <div>
      <UserAvatar src={src} alt="avatar" />
    </div>
  );
};

export default Avatar;
