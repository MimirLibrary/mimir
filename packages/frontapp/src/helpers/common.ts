export const shortenText = (text: string, numberToShorten: number) => {
  return text.trim().slice(0, numberToShorten);
};
