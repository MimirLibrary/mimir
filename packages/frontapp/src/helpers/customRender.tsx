import '@testing-library/jest-dom';
import { render, RenderOptions } from '@testing-library/react';
import React, { FC, ReactElement } from 'react';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../app/store';
import { ApolloProvider } from '@apollo/client';
import { client } from '@mimir/apollo-client';

jest.mock('react-i18next', () => ({
  // this mock makes sure any components using the translate hook can use it without a warning being shown
  useTranslation: () => {
    return {
      t: (str: string | null) => str,
      i18n: {
        changeLanguage: () =>
          new Promise(() => {
            return;
          }),
      },
    };
  },
}));

jest.unmock('react-redux');

const wrapper: FC<{ children: React.ReactNode }> = ({ children }) => {
  return (
    <BrowserRouter>
      <ApolloProvider client={client}>
        <Provider store={store}>{children}</Provider>
      </ApolloProvider>
    </BrowserRouter>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => {
  return render(ui, { wrapper, ...options });
};

export { customRender as render };
