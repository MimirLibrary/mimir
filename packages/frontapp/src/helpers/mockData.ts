export const mockStatus = [
  {
    person_id: 1,
    status: 'Free',
    created_at: '2022-09-08T03:58:39.228Z',
    id: 1,
    material_id: 1,
  },
];

export const mockMaterial = [
  {
    __typename: 'Material',
    author: 'test Author',
    category: 'test Category',
    created_at: '2022-09-08T03:58:39.228Z',
    description: 'Test Description',
    id: '1',
    id_type: 'Test Id_Type',
    identifier: '1111111111',
    is_donated: false,
    notification: [],
    picture: '',
    statuses: mockStatus,
    title: 'test',
    type: 'Book',
    updated_at: '2022-09-08T03:58:39.228Z',
  },
];
