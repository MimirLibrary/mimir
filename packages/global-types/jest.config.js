module.exports = {
  displayName: 'global-types',
  preset: '../../jest.preset.js',
  transform: {
    '^.+\\.[tj]sx?$': 'babel-jest',
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../coverage/packages/global-types',
};
