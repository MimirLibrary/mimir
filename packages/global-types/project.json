{
  "sourceRoot": "packages/global-types/src",
  "targets": {
    "build": {
      "executor": "@nrwl/js:tsc",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/packages/global-types",
        "main": "packages/global-types/src/index.ts",
        "tsConfig": "packages/global-types/tsconfig.lib.json",
        "assets": ["packages/global-types/*.md"]
      }
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["packages/global-types/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/packages/global-types"],
      "options": {
        "jestConfig": "packages/global-types/jest.config.js",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
