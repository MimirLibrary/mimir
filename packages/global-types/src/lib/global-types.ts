
/*
 * -------------------------------------------------------
 * THIS FILE WAS AUTOMATICALLY GENERATED (DO NOT MODIFY)
 * -------------------------------------------------------
 */

/* tslint:disable */
/* eslint-disable */
export enum Permissions {
    GRANT_REVOKE_MANAGER = "GRANT_REVOKE_MANAGER"
}

export interface CreateStateInput {
    person_id: number;
    state: boolean;
    description?: Nullable<string>;
}

export interface BookInput {
    identifier: string;
    person_id: number;
}

export interface ProlongTimeInput {
    person_id: number;
    material_id: number;
}

export interface CreateLocationInput {
    location: string;
}

export interface RemoveLocationInput {
    location_id: number;
}

export interface CreateMaterialInput {
    identifier: string;
    id_type: string;
    location_id: number;
    type: string;
    title: string;
    picture?: Nullable<string>;
    author: string;
    category: string;
    description: string;
}

export interface RemoveMaterialInput {
    identifier: string;
    location_id: number;
    type: string;
}

export interface UpdateMaterialInput {
    identifier?: Nullable<string>;
    id_type?: Nullable<string>;
    location_id?: Nullable<number>;
    type?: Nullable<string>;
    title?: Nullable<string>;
    picture?: Nullable<string>;
    author?: Nullable<string>;
    category?: Nullable<string>;
    updated_at: DateTime;
    description?: Nullable<string>;
}

export interface SearchOneMaterial {
    identifier: string;
    locations?: Nullable<number[]>;
}

export interface DonateBookInput {
    identifier: string;
    id_type: string;
    location_id: number;
    type: string;
    title: string;
    author: string;
    category: string;
    description?: Nullable<string>;
    picture?: Nullable<string>;
    person_id: number;
    role: string;
    is_donated: boolean;
}

export interface SearchInput {
    search?: Nullable<string>;
    locations?: Nullable<number[]>;
}

export interface CreateMessageInput {
    title: string;
    message: string;
    material_id?: Nullable<number>;
    person_id: number;
    location_id: number;
}

export interface CreateAnswerNotification {
    id: number;
    person_id: number;
    message: string;
}

export interface CreateNotificationInput {
    material_id: number;
    person_id: number;
}

export interface CreateSimpleNotification {
    person_id: number;
    message: string;
}

export interface RemoveNotificationInput {
    material_id: number;
    person_id: number;
}

export interface CreatePersonInput {
    smg_id: string;
    location_id: number;
    type: string;
}

export interface UpdatePersonLocationInput {
    person_id: number;
    location_id: number;
}

export interface CreateStatusInput {
    material_id: number;
    person_id: number;
    status: string;
}

export interface BlockedUsers {
    id: string;
    description?: Nullable<string>;
    state: boolean;
    created_at: DateTime;
    person: Person;
}

export interface IQuery {
    getBlocksByPerson(person_id: string): Nullable<Nullable<BlockedUsers>[]> | Promise<Nullable<Nullable<BlockedUsers>[]>>;
    getReasonOfBlock(person_id: string): Nullable<BlockedUsers> | Promise<Nullable<BlockedUsers>>;
    getAllTakenItems(person_id: number): Nullable<Status>[] | Promise<Nullable<Status>[]>;
    getItemsForClaimHistory(person_id: number): Nullable<Status>[] | Promise<Nullable<Status>[]>;
    getAllLocations(): Nullable<Location>[] | Promise<Nullable<Location>[]>;
    getAllMaterials(locations?: Nullable<number[]>, limit?: Nullable<number>, offset?: Nullable<number>): Nullable<Material>[] | Promise<Nullable<Material>[]>;
    getMaterialById(id: string): Material | Promise<Material>;
    searchOfMaterials(input: SearchInput): Nullable<Nullable<Material>[]> | Promise<Nullable<Nullable<Material>[]>>;
    getMaterialByIdentifier(input: SearchOneMaterial): Material | Promise<Material>;
    getMaterialByIdentifierFromMetadata(identifier: string): Nullable<IMetaOfMaterial> | Promise<Nullable<IMetaOfMaterial>>;
    getAllDonatedMaterialsByPerson(id: string): Nullable<Nullable<Material>[]> | Promise<Nullable<Nullable<Material>[]>>;
    getMessagesByPerson(person_id: string): Nullable<Nullable<Message>[]> | Promise<Nullable<Nullable<Message>[]>>;
    getAllMessages(location_id: number): Nullable<Message[]> | Promise<Nullable<Message[]>>;
    getNotificationsByPerson(person_id: number): Nullable<Notification>[] | Promise<Nullable<Notification>[]>;
    getNotificationsByMaterial(material_id: number): Nullable<Notification>[] | Promise<Nullable<Notification>[]>;
    getOnePerson(id: string): Person | Promise<Person>;
    getAllPersons(username?: Nullable<string>, locations?: Nullable<number[]>): Person[] | Promise<Person[]>;
    getStatusesByPerson(person_id: string): Nullable<Status>[] | Promise<Nullable<Status>[]>;
    getStatusesByMaterial(material_id: string): Nullable<Status>[] | Promise<Nullable<Status>[]>;
    getAllStatusesIsOverdue(locations?: Nullable<number[]>): Nullable<Status>[] | Promise<Nullable<Status>[]>;
    welcome(): string | Promise<string>;
}

export interface IMutation {
    createState(input: CreateStateInput): Nullable<BlockedUsers> | Promise<Nullable<BlockedUsers>>;
    claimBook(input?: Nullable<BookInput>): BookUnionResult | Promise<BookUnionResult>;
    returnItem(input?: Nullable<BookInput>): BookUnionResult | Promise<BookUnionResult>;
    rejectItem(input?: Nullable<BookInput>): BookUnionResult | Promise<BookUnionResult>;
    prolongClaimPeriod(input?: Nullable<ProlongTimeInput>): BookUnionResult | Promise<BookUnionResult>;
    createLocation(input: CreateLocationInput): Nullable<Location> | Promise<Nullable<Location>>;
    removeLocation(input: RemoveLocationInput): Nullable<Location> | Promise<Nullable<Location>>;
    createMaterial(input: CreateMaterialInput): Material | Promise<Material>;
    removeMaterial(input: RemoveMaterialInput): Material | Promise<Material>;
    updateMaterial(input: UpdateMaterialInput): Material | Promise<Material>;
    donateBook(input: DonateBookInput): Material | Promise<Material>;
    createMessageForManager(input: CreateMessageInput): Message | Promise<Message>;
    createNotification(input: CreateNotificationInput): Nullable<Notification> | Promise<Nullable<Notification>>;
    removeNotification(input: RemoveNotificationInput): Nullable<Notification> | Promise<Nullable<Notification>>;
    createAnswerNotification(input: CreateAnswerNotification): Notification | Promise<Notification>;
    createSimpleNotification(input?: Nullable<CreateSimpleNotification>): Notification | Promise<Notification>;
    createPerson(input: CreatePersonInput): Person | Promise<Person>;
    addPersonLocation(input: UpdatePersonLocationInput): Person | Promise<Person>;
    removePersonLocation(input: UpdatePersonLocationInput): Person | Promise<Person>;
    changePersonRole(person_id: number, type: string): Person | Promise<Person>;
    createStatus(input: CreateStatusInput): Status | Promise<Status>;
}

export interface Location {
    id: string;
    location: string;
    persons?: Nullable<Nullable<Person>[]>;
    materials?: Nullable<Nullable<Material>[]>;
}

export interface Material {
    id: string;
    identifier: string;
    id_type: string;
    type: string;
    location_id: number;
    created_at: DateTime;
    updated_at: DateTime;
    title: string;
    picture?: Nullable<string>;
    author: string;
    category: string;
    location: Location;
    statuses: Nullable<Status>[];
    notifications: Nullable<Notification>[];
    messages: Nullable<Message>[];
    description: string;
    is_donated: boolean;
}

export interface ResponseMetadata {
    idType: string;
    value: string;
}

export interface Author {
    id?: Nullable<string>;
    name?: Nullable<string>;
}

export interface Publisher {
    id?: Nullable<string>;
    name?: Nullable<string>;
}

export interface Meta {
    ageRestriction?: Nullable<string>;
    coverType?: Nullable<string>;
    dimensions?: Nullable<string>;
    manufacturer?: Nullable<string>;
    mass?: Nullable<string>;
    numberOfPages?: Nullable<string>;
    price?: Nullable<string>;
    series?: Nullable<string>;
    sku?: Nullable<string>;
}

export interface IMaterialMeta {
    authors?: Nullable<Nullable<Author>[]>;
    cover?: Nullable<string>;
    description?: Nullable<string>;
    title?: Nullable<string>;
    yearPublishedAt?: Nullable<number>;
    meta?: Nullable<Meta>;
    publisher?: Nullable<Publisher>;
}

export interface IMetaOfMaterial {
    idType?: Nullable<string>;
    value?: Nullable<string>;
    material?: Nullable<IMaterialMeta>;
}

export interface Message {
    id: string;
    material_id?: Nullable<number>;
    person_id: number;
    created_at: DateTime;
    person: Person;
    material: Material;
    title: string;
    message: string;
    location_id: number;
}

export interface Notification {
    id: string;
    material_id?: Nullable<number>;
    person_id: number;
    created_at: DateTime;
    person: Person;
    material?: Nullable<Material>;
    message?: Nullable<string>;
}

export interface Person {
    id: string;
    smg_id: string;
    position: string;
    type: string;
    username: string;
    email: string;
    avatar: string;
    created_at: DateTime;
    permissions?: Nullable<Nullable<Permissions>[]>;
    statuses?: Nullable<Nullable<Status>[]>;
    notifications?: Nullable<Nullable<Notification>[]>;
    location?: Nullable<Location[]>;
    messages?: Nullable<Nullable<Message>[]>;
    states?: Nullable<Nullable<BlockedUsers>[]>;
}

export interface Status {
    id: string;
    material_id: number;
    person_id: number;
    status: string;
    created_at: DateTime;
    material: Material;
    person: Person;
}

export interface Error {
    message: string;
}

export type DateTime = any;
export type Time = any;
export type Timestamp = any;
export type TimeZone = any;
export type UtcOffset = any;
export type Duration = any;
export type ISO8601Duration = any;
export type LocalDate = any;
export type LocalTime = any;
export type LocalEndTime = any;
export type EmailAddress = any;
export type NegativeFloat = any;
export type NegativeInt = any;
export type NonEmptyString = any;
export type NonNegativeFloat = any;
export type NonNegativeInt = any;
export type NonPositiveFloat = any;
export type NonPositiveInt = any;
export type PhoneNumber = any;
export type PositiveFloat = any;
export type PositiveInt = any;
export type PostalCode = any;
export type UnsignedFloat = any;
export type UnsignedInt = any;
export type URL = any;
export type BigInt = any;
export type Long = any;
export type Byte = any;
export type UUID = any;
export type GUID = any;
export type Hexadecimal = any;
export type HexColorCode = any;
export type HSL = any;
export type HSLA = any;
export type IPv4 = any;
export type IPv6 = any;
export type ISBN = any;
export type JWT = any;
export type Latitude = any;
export type Longitude = any;
export type MAC = any;
export type Port = any;
export type RGB = any;
export type RGBA = any;
export type SafeInt = any;
export type USCurrency = any;
export type Currency = any;
export type JSON = any;
export type JSONObject = any;
export type IBAN = any;
export type ObjectID = any;
export type Void = any;
export type DID = any;
export type CountryCode = any;
export type Locale = any;
export type RoutingNumber = any;
export type AccountNumber = any;
export type BookUnionResult = Status | Error;
type Nullable<T> = T | null;
