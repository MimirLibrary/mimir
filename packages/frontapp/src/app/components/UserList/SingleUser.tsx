import React, { FC } from 'react';
import Avatar from '../Avatar';
import { mockData } from './mockData';
import styled from '@emotion/styled';
import { colors, dimensions } from '@mimir/ui-kit';
import { useNavigate } from 'react-router-dom';
import { IClaimHistory } from '../../models/helperFunctions/claimHistory';
import ClaimHistory from '../ClaimHistory';
import { RoutesTypes } from '../../../utils/routes';

const InfoWrapper = styled.div`
  display: flex;
  flex-direction: column;
  margin-left: ${dimensions.base};
  row-gap: ${dimensions.xs_2};
  > p:first-of-type {
    margin-bottom: ${dimensions.xs_2};
  }
`;

const ClaimsWrapper = styled.div`
  font-size: ${dimensions.sm};
  row-gap: 4px;
`;

const CardWrapper = styled.div`
  display: flex;
  flex-direction: row;
  background: ${colors.bg_secondary};
  width: 320px;
  margin: 0 0 ${dimensions.xl};
  height: 151px;
  padding: ${dimensions.xl_2};
  border-radius: ${dimensions.xs_1};
  transition: box-shadow 0.3s;
  cursor: pointer;
  :hover {
    box-shadow: 0px 6px 14px -6px rgba(24, 39, 75, 0.08),
      0px 10px 32px -4px rgba(24, 39, 75, 0.08);
  }
`;
const AvatarWrapper = styled.div`
  height: 71px;
  width: 71px;
  border-radius: 50%;
`;

const Description = styled.p`
  font-weight: 500;
  font-size: ${dimensions.base};
  line-height: ${dimensions.xl};
`;

export interface ISingleUser {
  avatar: string | null | undefined;
  id: string;
  name: string;
  statuses: IClaimHistory[];
  className?: string;
}

const SingleUser: FC<ISingleUser> = ({
  id = '',
  statuses = [],
  name,
  avatar,
  className,
}) => {
  const navigate = useNavigate();
  const handleUserRedirect = () => {
    navigate(`${RoutesTypes.READERS}/${id}`);
  };

  return (
    <CardWrapper
      className={className}
      onClick={handleUserRedirect}
      data-testid="single-user"
    >
      <AvatarWrapper>
        <Avatar src={avatar || mockData.avatar} />
      </AvatarWrapper>
      <InfoWrapper>
        <Description>{name}</Description>
        <ClaimsWrapper>
          <ClaimHistory statuses={statuses} />
        </ClaimsWrapper>
      </InfoWrapper>
    </CardWrapper>
  );
};

export default SingleUser;
