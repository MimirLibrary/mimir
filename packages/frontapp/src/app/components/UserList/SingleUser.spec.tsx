import { render } from '../../../helpers/customRender';
import SingleUser from './SingleUser';
import { StatusTypes } from '@mimir/global-types';
import { screen } from '@testing-library/react';
import Avatar from '../../../assets/avatar.jpg';

describe('render SingleUserCard component', () => {
  it('should be correctly render', () => {
    render(
      <SingleUser
        avatar=""
        id="1"
        statuses={[
          {
            status: StatusTypes.FREE,
            material_id: 13,
            created_at: new Date('2022-09-21T08:33:02.239Z'),
          },
        ]}
        name="Test"
      />
    );

    expect(screen.getByTestId('single-user')).toBeInTheDocument();
    expect(screen.getByText('Test')).toBeInTheDocument();
    expect(screen.getByRole('img')).toHaveAttribute('src', Avatar);
  });

  it('should be correctly render without name', () => {
    render(
      <SingleUser
        avatar=""
        id="1"
        statuses={[
          {
            status: StatusTypes.FREE,
            material_id: 13,
            created_at: new Date('2022-09-21T08:33:02.239Z'),
          },
        ]}
        name=""
      />
    );

    expect(screen.getByTestId('single-user')).toBeInTheDocument();
    expect(screen.queryByText('Test')).not.toBeInTheDocument();
    expect(screen.getByRole('img')).toHaveAttribute('src', Avatar);
  });

  it('renders correctly, with snapshots', () => {
    render(
      <SingleUser
        avatar={Avatar}
        id="1"
        statuses={[
          {
            status: StatusTypes.FREE,
            material_id: 13,
            created_at: new Date('2022-09-21T08:33:02.239Z'),
          },
        ]}
        name="Test"
      />
    );
    expect(screen.getByTestId('single-user')).toMatchSnapshot();
  });
});
