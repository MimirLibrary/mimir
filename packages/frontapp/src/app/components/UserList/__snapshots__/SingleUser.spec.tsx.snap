// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render SingleUserCard component renders correctly, with snapshots 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  background: #FFFFFF;
  width: 320px;
  margin: 0 0 1.25rem;
  height: 151px;
  padding: 1.5rem;
  border-radius: 0.625rem;
  -webkit-transition: box-shadow 0.3s;
  transition: box-shadow 0.3s;
  cursor: pointer;
}

.emotion-0:hover {
  box-shadow: 0px 6px 14px -6px rgba(24, 39, 75, 0.08),0px 10px 32px -4px rgba(24, 39, 75, 0.08);
}

.emotion-1 {
  height: 71px;
  width: 71px;
  border-radius: 50%;
}

.emotion-2 {
  max-height: 7rem;
  max-width: 7rem;
  width: 100%;
  height: 100%;
  border-radius: 50%;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-left: 1rem;
  row-gap: 0.5rem;
}

.emotion-3>p:first-of-type {
  margin-bottom: 0.5rem;
}

.emotion-4 {
  font-weight: 500;
  font-size: 1rem;
  line-height: 1.25rem;
}

.emotion-5 {
  font-size: 0.875rem;
  row-gap: 4px;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  row-gap: inherit;
}

.emotion-7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-column-gap: 4px;
  column-gap: 4px;
}

.emotion-8 {
  font-weight: 500;
  font-size: inherit;
  line-height: 1.25rem;
}

.emotion-9 {
  font-weight: 300;
  font-size: inherit;
  line-height: 1.25rem;
}

<div
  class="emotion-0"
  data-testid="single-user"
>
  <div
    class="emotion-1"
  >
    <div>
      <img
        alt="avatar"
        class="emotion-2"
        src="avatar.jpg"
      />
    </div>
  </div>
  <div
    class="emotion-3"
  >
    <p
      class="emotion-4"
    >
      Test
    </p>
    <div
      class="emotion-5"
    >
      <div
        class="emotion-6"
      >
        <div
          class="emotion-7"
        >
          <p
            class="emotion-8"
          />
          <p
            class="emotion-9"
          >
            -
          </p>
        </div>
        <div
          class="emotion-7"
        >
          <p
            class="emotion-8"
          />
          <p
            class="emotion-9"
          >
            -
          </p>
        </div>
        <div
          class="emotion-7"
        >
          <p
            class="emotion-8"
          >
            undefined: 
          </p>
          <p
            class="emotion-9"
          >
            -
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
